File: src/detection/os/os_apple.c
#include "os.h"
#include "common/properties.h"
#include "common/sysctl.h"

#include <stdlib.h>
#include <string.h>

typedef enum PListKey
{
    PLIST_KEY_NAME,
    PLIST_KEY_VERSION,
    PLIST_KEY_BUILD,
    PLIST_KEY_OTHER
} PListKey;

static void parseSystemVersion(FFOSResult* os)
{
    FILE* plist = fopen("/System/Library/CoreServices/SystemVersion.plist", "r");
    if(plist == NULL)
    return;

    char* line = NULL;
    size_t len = 0;
    PListKey key = PLIST_KEY_OTHER;

    FFstrbuf keyBuffer;
    ffStrbufInit(&keyBuffer);

    while(getline(&line, &len, plist) != EOF)
    {
        if(ffParsePropLine(line, "<key>", &keyBuffer))
        {
            if(ffStrbufIgnCaseCompS(&keyBuffer, "ProductName") == 0)
                key = PLIST_KEY_NAME;
            else if(ffStrbufIgnCaseCompS(&keyBuffer, "ProductUserVisibleVersion") == 0)
                key = PLIST_KEY_VERSION;
            else if(ffStrbufIgnCaseCompS(&keyBuffer, "ProductBuildVersion") == 0)
                key = PLIST_KEY_BUILD;
            else
                key = PLIST_KEY_OTHER;

            ffStrbufClear(&keyBuffer);
            continue;
        }

        if(key == PLIST_KEY_NAME)
            ffParsePropLine(line, "<string>", &os->name);
        else if(key == PLIST_KEY_VERSION)
            ffParsePropLine(line, "<string>", &os->version);
        else if(key == PLIST_KEY_BUILD)
            ffParsePropLine(line, "<string>", &os->buildID);
    }

    ffStrbufDestroy(&keyBuffer);

    if(line != NULL)
        free(line);

    fclose(plist);
}

void parseOSXSoftwareLicense(FFOSResult* os)
{
    FILE* rtf = fopen("/System/Library/CoreServices/Setup Assistant.app/Contents/Resources/en.lproj/OSXSoftwareLicense.rtf", "r");
    if(rtf == NULL)
        return;

    char* line = NULL;
    size_t len = 0;
    const char* searchStr = "\\f0\\b SOFTWARE LICENSE AGREEMENT FOR macOS ";
    const size_t searchLen = strlen(searchStr);
    while(getline(&line, &len, rtf) != EOF)
    {
        if (strncmp(line, searchStr, searchLen) == 0)
        {
            ffStrbufAppendS(&os->codename, line + searchLen);
            ffStrbufTrimRight(&os->codename, '\n');
            ffStrbufTrimRight(&os->codename, '\\');
            break;
        }
    }

    if(line != NULL)
        free(line);

    fclose(rtf);
}

void ffDetectOSImpl(FFOSResult* os, const FFinstance* instance)
{
    FF_UNUSED(instance);

    ffStrbufInit(&os->name);
    ffStrbufInit(&os->version);
    ffStrbufInit(&os->buildID);
    ffStrbufInit(&os->id);
    ffStrbufInit(&os->prettyName);
    ffStrbufInit(&os->versionID);
    ffStrbufInit(&os->systemName);
    ffStrbufInit(&os->architecture);

    ffStrbufInitA(&os->codename, 0);
    ffStrbufInitA(&os->idLike, 0);
    ffStrbufInitA(&os->variant, 0);
    ffStrbufInitA(&os->variantID, 0);

    parseSystemVersion(os);

    if(ffStrbufStartsWithIgnCaseS(&os->name, "MacOS"))
        ffStrbufAppendS(&os->id, "macos");

    if(os->version.length == 0)
        ffSysctlGetString("kern.osproductversion", &os->version);

    if(os->buildID.length == 0)
        ffSysctlGetString("kern.osversion", &os->buildID);

    ffStrbufAppend(&os->prettyName, &os->name);
    ffStrbufAppend(&os->versionID, &os->version);
    ffSysctlGetString("kern.ostype", &os->systemName);
    ffSysctlGetString("hw.machine", &os->architecture);

    parseOSXSoftwareLicense(os);
}

File: src/detection/os/os_apple.m
#include "os.h"
#include "common/io/io.h"
#include "common/sysctl.h"
#include "util/stringUtils.h"
#include "util/mallocHelper.h"

#include <stdlib.h>
#include <string.h>
#import <Foundation/Foundation.h>

static bool parseSystemVersion(FFOSResult* os)
{
    NSError* error;
    NSString* fileName = @"file:///System/Library/CoreServices/SystemVersion.plist";
    NSDictionary* dict = [NSDictionary dictionaryWithContentsOfURL:[NSURL URLWithString:fileName]
                                       error:&error];
    if(error)
        return false;

    NSString* value;

    if((value = dict[@"ProductName"]))
        ffStrbufInitS(&os->name, value.UTF8String);
    if((value = dict[@"ProductUserVisibleVersion"]))
        ffStrbufInitS(&os->version, value.UTF8String);
    if((value = dict[@"ProductBuildVersion"]))
        ffStrbufInitS(&os->buildID, value.UTF8String);
    if (ffStrbufStartsWithS(&os->version, "16."))
    {
        // macOS 26 Tahoe. #1809
        os->version.chars[0] = '2';
    }

    return true;
}

static bool detectOSCodeName(FFOSResult* os)
{
    // https://en.wikipedia.org/wiki/MacOS_version_history
    char* str_end;
    const char* version = os->version.chars;
    unsigned long num = strtoul(version, &str_end, 10);
    if (str_end == version) return false;

    switch (num)
    {
        case 26:
        case 16: ffStrbufSetStatic(&os->codename, "Tahoe"); return true;
        case 15: ffStrbufSetStatic(&os->codename, "Sequoia"); return true;
        case 14: ffStrbufSetStatic(&os->codename, "Sonoma"); return true;
        case 13: ffStrbufSetStatic(&os->codename, "Ventura"); return true;
        case 12: ffStrbufSetStatic(&os->codename, "Monterey"); return true;
        case 11: ffStrbufSetStatic(&os->codename, "Big Sur"); return true;
        case 10: {
            version = str_end + 1;
            num = strtoul(version, &str_end, 10);
            if (str_end == version) return false;

            switch (num)
            {
                case 16: ffStrbufSetStatic(&os->codename, "Big Sur"); return true;
                case 15: ffStrbufSetStatic(&os->codename, "Catalina"); return true;
                case 14: ffStrbufSetStatic(&os->codename, "Mojave"); return true;
                case 13: ffStrbufSetStatic(&os->codename, "High Sierra"); return true;
                case 12: ffStrbufSetStatic(&os->codename, "Sierra"); return true;
                case 11: ffStrbufSetStatic(&os->codename, "El Capitan"); return true;
                case 10: ffStrbufSetStatic(&os->codename, "Yosemite"); return true;
                case 9: ffStrbufSetStatic(&os->codename, "Mavericks"); return true;
                case 8: ffStrbufSetStatic(&os->codename, "Mountain Lion"); return true;
                case 7: ffStrbufSetStatic(&os->codename, "Lion"); return true;
                case 6: ffStrbufSetStatic(&os->codename, "Snow Leopard"); return true;
                case 5: ffStrbufSetStatic(&os->codename, "Leopard"); return true;
                case 4: ffStrbufSetStatic(&os->codename, "Tiger"); return true;
                case 3: ffStrbufSetStatic(&os->codename, "Panther"); return true;
                case 2: ffStrbufSetStatic(&os->codename, "Jaguar"); return true;
                case 1: ffStrbufSetStatic(&os->codename, "Puma"); return true;
                case 0: ffStrbufSetStatic(&os->codename, "Cheetah"); return true;
            }
        }
    }

    return false;
}

void ffDetectOSImpl(FFOSResult* os)
{
    parseSystemVersion(os);

    ffStrbufSetStatic(&os->id, "macos");

    if(__builtin_expect(os->name.length == 0, 0))
        ffStrbufSetStatic(&os->name, "macOS");

    if(__builtin_expect(os->version.length == 0, 0))
        ffSysctlGetString("kern.osproductversion", &os->version);

    if(__builtin_expect(os->buildID.length == 0, 0))
        ffSysctlGetString("kern.osversion", &os->buildID);

    ffStrbufAppend(&os->versionID, &os->version);

    detectOSCodeName(os);

    ffStrbufSetF(&os->prettyName, "%s %s %s (%s)", os->name.chars, os->codename.chars, os->version.chars, os->buildID.chars);
}
